drv/cosmo-hf/src/main.rs

```diff
@@ -327,6 +341,36 @@ impl FlashDriver {
         self.wait_fpga_busy();
     }
 
+    fn get_flash_mux_state(&self) -> drv_hf_api::HfMuxState {
+        let v = self.read_reg(reg::SPICR);
+        if v & reg::spicr::SP5_OWNS_FLASH != 0 {
+            drv_hf_api::HfMuxState::HostCPU
+        } else {
+            drv_hf_api::HfMuxState::SP
+        }
+    }
+
+    /// Returns an error if the flash mux state is not `HfMuxState::SP`
+    fn check_flash_mux_state(&self) -> Result<(), drv_hf_api::HfError> {
+        match self.get_flash_mux_state() {
+            drv_hf_api::HfMuxState::SP => Ok(()),
+            drv_hf_api::HfMuxState::HostCPU => {
+                Err(drv_hf_api::HfError::NotMuxedToSP)
+            }
+        }
+    }
+
+    fn set_flash_mux_state(&self, ms: drv_hf_api::HfMuxState) {
+        match ms {
```

Comment:
**suggestion:** If `modify_reg` took `F: FnOnce(u32)` then (I believe) you could match on the HfMuxState to get the right side of the bitwise or, and write the rest of the function as a single `modify_reg` that uses the result of that match.

I didn't check if the compiler output is the same either way, but the code might be clearer that way: There wouldn't be two mostly equal function call paths with just one load-bearing exclamation mark hidden somewhere in there.

Reply:
Hmm, I think you could do this refactoring even with the current type for `modify_reg` -- and I bet it would generate less code, just because the way things are currently written, each `modify_reg` is a monomorphization site (due to the closure types being disjoint) and that tends to encourage more code generation.

Reply:
Done in c2461ba96956a5c7ff08cd153f3cb46c5713ab29